File: engine/src/main/java/top/skyeyefast/mchjong/engine/ReplayPlayback.java
package top.skyeyefast.mchjong.engine;

import java.util.ArrayList;
import java.util.List;

/** A read-only timeline. It never sends game actions or exposes any still-playing hand. */
public final class ReplayPlayback {
    private ReplayPlayback() {}
    public record Frame(List<TableView.Seat> seats, List<Integer> dora, int cursor, int steps,
                        boolean settled, ReplayHand.Event event) {
        public Frame { seats = List.copyOf(seats); dora = List.copyOf(dora); }
    }

    private static final class Seat {
        final List<Integer> hand;
        final List<Meld> melds = new ArrayList<>();
        final List<Discard> river = new ArrayList<>();
        final List<Integer> norths = new ArrayList<>();
        int points, drawn = Tile.ABSENT;
        boolean riichi, nextSideways;
        Seat(List<Integer> hand, int points) { this.hand = new ArrayList<>(hand); this.points = points; }
        void remove(int tile) {
            if (!hand.remove(Integer.valueOf(tile))) throw new IllegalArgumentException("Replay uses an unowned tile");
        }
    }

    public static Frame at(ReplayMatch match, int handIndex, int requested) {
        ReplayHand hand = match.hands().get(handIndex);
        int steps = hand.events().size() + 1;
        int cursor = Math.clamp(requested, 0, steps);
        if (cursor == steps) return new Frame(hand.finalSeats(), hand.dora(), cursor, steps, true, null);
        var seats = new ArrayList<Seat>();
        for (int i = 0; i < match.rules().players(); i++) seats.add(new Seat(hand.initialHands().get(i), hand.initialPoints().get(i)));
        var dora = new ArrayList<>(hand.initialDora());
        for (int i = 0; i < cursor; i++) {
            ReplayHand.Event event = hand.events().get(i);
            if (event.kind() == ReplayHand.Kind.DORA) { dora.add(event.tile()); continue; }
            Seat seat = seats.get(event.seat());
            switch (event.kind()) {
                case DRAW -> {
                    if (seat.hand.contains(event.tile()) || seat.hand.size() >= 14)
                        throw new IllegalArgumentException("Invalid replay draw");
                    seat.hand.add(event.tile()); seat.drawn = event.tile();
                }
                case DISCARD -> {
                    seat.remove(event.tile());
                    seat.drawn = Tile.ABSENT;
                    seat.river.add(new Discard(event.tile(), event.riichi() || seat.nextSideways, false, event.tsumogiri()));
                    seat.nextSideways = false;
                }
                case RIICHI -> { seat.riichi = true; seat.points -= 1000; }
                case NUKI -> {
                    if (event.committed()) {
                        seat.remove(event.tile()); seat.norths.add(event.tile()); seat.drawn = Tile.ABSENT;
                    }
                }
                case MELD -> {
                    if (!event.committed()) continue;
                    Meld meld = event.meld();
                    if (meld.type() == Meld.Type.ADDED_KAN) {
                        seat.remove(event.tile());
                        int pon = -1;
                        for (int m = 0; m < seat.melds.size(); m++)
                            if (seat.melds.get(m).type() == Meld.Type.PON && seat.melds.get(m).kind() == meld.kind()) pon = m;
                        if (pon < 0) throw new IllegalArgumentException("Added kan has no pon");
                        seat.melds.set(pon, meld);
                    } else {
                        for (int tile : meld.tiles()) if (meld.closed() || tile != meld.calledTile()) seat.remove(tile);
                        seat.melds.add(meld);
                        if (!meld.closed()) {
                            Seat source = seats.get(meld.fromSeat());
                            if (source.river.isEmpty() || source.river.getLast().tile() != meld.calledTile())
                                throw new IllegalArgumentException("Meld does not claim the last discard");
                            Discard discard = source.river.getLast();
                            source.river.set(source.river.size() - 1, discard.markCalled());
                            source.nextSideways = discard.riichi();
                        }
                    }
                    seat.drawn = Tile.ABSENT;
                }
                default -> throw new IllegalStateException("Unexpected replay event");
            }
        }
        var view = new ArrayList<TableView.Seat>();
        for (int i = 0; i < seats.size(); i++) {
            Seat seat = seats.get(i);
            seat.hand.sort(Tile.ORDER);
            if (seat.drawn >= 0 && seat.hand.remove(Integer.valueOf(seat.drawn))) seat.hand.add(seat.drawn);
            var player = match.participants().get(i);
            view.add(new TableView.Seat(player.name(), true, player.bot(), false, seat.points, seat.hand, seat.drawn,
                seat.melds, seat.river, seat.norths, seat.riichi, true));
        }
        return new Frame(view, dora, cursor, steps, false, cursor == 0 ? null : hand.events().get(cursor - 1));
    }
}

File: engine/src/main/kotlin/top/skyeyefast/mchjong/engine/ReplayPlayback.kt
package top.skyeyefast.mchjong.engine

/** A read-only timeline. It never sends game actions or exposes any still-playing hand. */
object ReplayPlayback {
    @JvmRecord
    data class Frame(
        val seats: List<TableView.Seat>,
        val dora: List<Int>,
        val cursor: Int,
        val steps: Int,
        val rawCursor: Int,
        val settled: Boolean,
        val event: ReplayHand.Event?,
    )

    @JvmRecord
    data class Timeline(val frames: List<Frame>) {
        init {
            require(frames.size >= 2 && frames.first().cursor == 0 && frames.last().settled) { "Invalid replay timeline" }
        }
    }

    private class Seat(hand: List<Int>, var points: Int) {
        val hand = hand.toMutableList()
        val melds = mutableListOf<Meld>()
        val river = mutableListOf<Discard>()
        val norths = mutableListOf<Int>()
        var drawn = Tile.ABSENT
        var riichi = false
        var nextSideways = false

        fun remove(tile: Int) {
            require(hand.remove(tile)) { "Replay uses an unowned tile" }
        }
    }

    @JvmStatic
    fun at(match: ReplayMatch, handIndex: Int, requested: Int): Frame {
        val hand = match.hands[handIndex]
        val steps = hand.events.size + 1
        val cursor = requested.coerceIn(0, steps)
        if (cursor == steps) {
            return Frame(java.util.List.copyOf(hand.finalSeats), java.util.List.copyOf(hand.dora), cursor, steps, hand.events.size, true, null)
        }
        val seats = MutableList(match.rules.players()) { Seat(hand.initialHands[it], hand.initialPoints[it]) }
        val dora = hand.initialDora.toMutableList()
        for (i in 0 until cursor) {
            val event = hand.events[i]
            if (event.kind == ReplayHand.Kind.DORA) {
                dora += event.tile
                continue
            }
            val seat = seats[event.seat]
            when (event.kind) {
                ReplayHand.Kind.DRAW -> {
                    require(event.tile !in seat.hand && seat.hand.size < 14) { "Invalid replay draw" }
                    seat.hand += event.tile
                    seat.drawn = event.tile
                }
                ReplayHand.Kind.DISCARD -> {
                    seat.remove(event.tile)
                    seat.drawn = Tile.ABSENT
                    seat.river += Discard(event.tile, event.riichi || seat.nextSideways, false, event.tsumogiri)
                    seat.nextSideways = false
                }
                ReplayHand.Kind.RIICHI -> {
                    seat.riichi = true
                    seat.points -= 1000
                }
                ReplayHand.Kind.NUKI -> if (event.committed) {
                    seat.remove(event.tile)
                    seat.norths += event.tile
                    seat.drawn = Tile.ABSENT
                }
                ReplayHand.Kind.MELD -> {
                    if (!event.committed) continue
                    val meld = event.meld!!
                    if (meld.type() == Meld.Type.ADDED_KAN) {
                        seat.remove(event.tile)
                        val pon = seat.melds.indexOfLast { it.type() == Meld.Type.PON && it.kind() == meld.kind() }
                        require(pon >= 0) { "Added kan has no pon" }
                        seat.melds[pon] = meld
                    } else {
                        for (tile in meld.tiles()) if (meld.closed() || tile != meld.calledTile()) seat.remove(tile)
                        seat.melds += meld
                        if (!meld.closed()) {
                            val source = seats[meld.fromSeat()]
                            require(source.river.isNotEmpty() && source.river.last().tile() == meld.calledTile()) {
                                "Meld does not claim the last discard"
                            }
                            val discard = source.river.last()
                            source.river[source.river.lastIndex] = discard.markCalled()
                            source.nextSideways = discard.riichi()
                        }
                    }
                    seat.drawn = Tile.ABSENT
                }
                ReplayHand.Kind.DORA -> error("Unexpected replay event")
            }
        }
        val view = seats.mapIndexed { i, seat ->
            seat.hand.sortWith(Tile.ORDER)
            if (seat.drawn >= 0 && seat.hand.remove(seat.drawn)) seat.hand += seat.drawn
            val player = match.participants[i]
            TableView.Seat(
                hand.finalSeats[i].entityBot(),
                player.name,
                true,
                player.bot,
                false,
                seat.points,
                seat.hand,
                seat.drawn,
                seat.melds,
                seat.river,
                seat.norths,
                seat.riichi,
                true,
                seat.riichi && hand.finalSeats[i].doubleRiichi(),
            )
        }
        return Frame(
            java.util.List.copyOf(view),
            java.util.List.copyOf(dora),
            cursor,
            steps,
            cursor,
            false,
            if (cursor == 0) null else hand.events[cursor - 1],
        )
    }

    /**
     * Compiles raw recorder events into user-facing replay steps. Riichi payment and dora reveal events are folded into
     * the action immediately before them, so navigation follows meaningful table actions instead of recorder internals.
     */
    @JvmStatic
    fun timeline(match: ReplayMatch, handIndex: Int): Timeline {
        val hand = match.hands[handIndex]
        val frames = mutableListOf(at(match, handIndex, 0))
        var eventIndex = 0
        while (eventIndex < hand.events.size) {
            val event = hand.events[eventIndex]
            if (event.kind == ReplayHand.Kind.RIICHI || event.kind == ReplayHand.Kind.DORA) {
                eventIndex++
                continue
            }
            var rawCursor = eventIndex + 1
            while (rawCursor < hand.events.size &&
                (hand.events[rawCursor].kind == ReplayHand.Kind.RIICHI || hand.events[rawCursor].kind == ReplayHand.Kind.DORA)
            ) rawCursor++
            frames += at(match, handIndex, rawCursor).copy(event = event)
            eventIndex = rawCursor
        }
        frames += at(match, handIndex, Int.MAX_VALUE)
        val last = frames.lastIndex
        return Timeline(java.util.List.copyOf(frames.mapIndexed { index, frame -> frame.copy(cursor = index, steps = last) }))
    }
}
